### PR TITLE
Add `Verifier` for `ReportData`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/1" target="_blank">Graphite</a> 
    * **PR #2** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/2" target="_blank">Graphite</a> 
      * **PR #3** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/3" target="_blank">Graphite</a> 
        * **PR #9** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/9" target="_blank">Graphite</a> 
          * **PR #10** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/10" target="_blank">Graphite</a> 
            * **PR #11** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/11" target="_blank">Graphite</a> 
              * **PR #12** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/12" target="_blank">Graphite</a> 
              * **PR #18** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/18" target="_blank">Graphite</a> 
                * **PR #20** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/20" target="_blank">Graphite</a> 
                  * **PR #21** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/21" target="_blank">Graphite</a> 
                    * **PR #22** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/22" target="_blank">Graphite</a> 
                      * **PR #23** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/23" target="_blank">Graphite</a> 
                        * **PR #24** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/24" target="_blank">Graphite</a> 
                          * **PR #25** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/25" target="_blank">Graphite</a>  👈
    * **PR #4** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/4" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/attestation/25?utm_source=stack-comment).

File: verifier/src/report_body.rs
Comment (NotGyro) [inline verifier/src/report_body.rs:345]: It might make sense to make this constant-time before merging the PR, although I don't know what would be involved in implementing a CT comparison on this type.

**Review by NotGyro — CHANGES_REQUESTED** [review]

File: verifier/src/report_body.rs
Comment (nick-mobilecoin) [reply verifier/src/report_body.rs:345]: Good call out!

I created #27 to track this, currently it's blocked on https://github.com/mobilecoinfoundation/sgx/issues/280

File: verifier/src/report_body.rs
Comment (samdealy) [inline verifier/src/report_body.rs:329]: Nit: the comment doesn't add any clarifying details so would recommend removing the parameter callout altogether.

**Review by samdealy — APPROVED** [review]

**Review by jcape — APPROVED** [review]
This is fine for now, but we should also add a `ReportDataMaskVerifier` for compatibility with EPID's sake.

Right now in EPID land, `report_data` contains two things: 32 bytes of static noise identity public key, and 32 bytes of app-specific data we want to attest to (e.g. in consensus, this is a block signing public key, in fog ingest this, is the ingress public key---i.e. the key that senders should encrypt their fog hint for). When we switch to DCAP, I'd like to change this to some kind of a MAC over an associated data structure that contains these keys (i.e. give us more than 64 bytes worth of data that we can attest to), but that's not how it works now.

The upshot is that right now, `mc-attest-ake` needs to be able to check that the first 32 bytes match, and ignore the rest of `ReportData`.

**Comment by nick-mobilecoin** [comment]
created #28 to address the partial verification

File: verifier/src/report_body.rs
Comment (nick-mobilecoin) [reply verifier/src/report_body.rs]: Thanks!
Bad habit from document all the things even if when the names match.  I can't find rust official docs, but it does seem rust prefers "clear" argument names over arguments being documented like this suggests https://www.tangramvision.com/blog/making-great-docs-with-rustdoc#-arguments

**Comment by nick-mobilecoin** [comment]
@samdealy bumped for re-review due to argument docs cleanup

**Review by samdealy — APPROVED** [review]

**Review by NotGyro — APPROVED** [review]
Looks good to me, especially now that https://github.com/mobilecoinfoundation/attestation/issues/27 is noted and won't get forgotten

**Pull request merged**
